### PR TITLE
Perception Encoder Integration

### PR DESCRIPTION
**Pull request by jacobsela**
## What changes are proposed in this pull request?

Starting an integration for [Perception Encoder](https://github.com/facebookresearch/perception_models/tree/main/apps/pe).

Only doing the ViT encoder of it now, can follow up with the rest in future PRs.

Install from the fork: https://github.com/voxel51/perception-models

I recommend running a visualization with this model on `foz.load_zoo_dataset("imagenet-sample")`

## How is this patch tested? If it is not, please explain why.
```python
import torch
import numpy as np

import fiftyone as fo
import fiftyone.zoo as foz
import fiftyone.utils.pe as fope

dataset = foz.load_zoo_dataset("quickstart", persistent=True)

model = foz.load_zoo_model(
    "PE-Core-L14-336-Vision-Encoder", device="cuda:7", project=True,
)

dataset.compute_embeddings(
    model,
    "pe_integration_test",
    batch_size=32,
    num_workers=8,
)

samples = dataset[:10]

embeddings = np.array(samples.values("pe_integration_test"))

import perception_models.core.vision_encoder.pe as pe
import perception_models.core.vision_encoder.transforms as pet
from PIL import Image

model = pe.VisionTransformer.from_config(
    name="PE-Core-L14-336", pretrained=True
).to("cuda:7")
transform = pet.get_image_transform(image_size=532)

filepaths = samples.values("filepath")
images = [transform(Image.open(fp)).unsqueeze(0).to("cuda:7") for fp in filepaths]
images = torch.cat(images, dim=0)
with torch.no_grad():
    outputs = model(images).cpu().numpy()

passed = np.allclose(embeddings, outputs, rtol=1e-5, atol=1e-5)
print("PE integration test passed!" if passed else "PE integration test failed!")

diff = np.abs(embeddings - outputs)
print(f"diff max = {diff.max()}")
print(f"diff mean = {np.mean(diff)}")
print(f"diff std = {np.std(diff)}")
```

prints:
```
PE integration test passed!
diff max = 1.9073486328125e-06
diff mean = 1.7749498226748983e-07
diff std = 1.479596249964743e-07
```

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

Added an integration with Perception Encoder (only ViT encoder for now).

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [x] Other

## Summary by CodeRabbit

* **New Features**
  * Adds Perception Encoder support to load and run Meta vision models within FiftyOne.
  * Enables automatic checkpoint download from HuggingFace and sensible defaults for checkpoint paths and image transforms.
  * Provides configurable inference options — pooling, projection, normalization, token handling — and flexible layer selection for generating embeddings.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: fiftyone/utils/pe.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+def download_pe_checkpoint(name: str, checkpoint_path: str):
+    path = f"hf://facebook/{name}:{name}.pt"
+
+    # Load from huggingface
+    path = path[len("hf://") :]
+    repo, file = path.split(":")
+
+    local_dir = checkpoint_path.rsplit("/", 1)[0]
+
+    return hf_hub_download(repo_id=repo, filename=file, local_dir=local_dir)
```
[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/utils/pe.py:29]: resolved

File: fiftyone/utils/pe.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+            if self.transforms_fcn == pe_transforms.get_image_transform:
+                # if the user has not specified otherwise, batches should not be ragged
+                self.ragged_batches = False
+
+        self.pool = self.parse_bool(d, "pool", True)
+        self.project = self.parse_bool(d, "project", True)
+        self.norm = self.parse_bool(d, "norm", True)
+        self.layer_idx = self.parse_int(d, "layer_idx", -1)
+        self.strip_cls_token = self.parse_bool(d, "strip_cls_token", False)
+
```
[the remark this answers is not in the document]
Comment (jacobsela) [reply fiftyone/utils/pe.py:116]: outdated

File: fiftyone/utils/pe.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        self.as_feature_extractor = True
+        self.layer_name = None
+
+        # defaults
+        # TODO: move to manifest
+        if self.entrypoint_fcn is None:
+            self.entrypoint_fcn = pe.VisionTransformer.from_config
+        if self.entrypoint_args is None:
+            self.entrypoint_args = {}
+        default_entrypoint_args = {
```
Comment (kaixi-wang) [inline fiftyone/utils/pe.py:88]: why are there "hardcoded" default args here and in the manifest https://github.com/voxel51/fiftyone/pull/6595/files#diff-d8412f633c4d9cc26284a55cf1ed0885f26efef3e3eca7997b3d6ad5b8d985adR8422

File: fiftyone/utils/pe.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+    def _forward_pass(self, imgs):
+        outputs = self._model.forward_features(
+            imgs,
+            norm=self.config.norm or self.config.project or self.config.pool,
+            layer_idx=self.config.layer_idx,
+            strip_cls_token=self.config.strip_cls_token,
+        )
+
+        if self.config.project or self.config.pool:
```
Comment (kaixi-wang) [inline fiftyone/utils/pe.py:129]: is this intentional? (if not self.config.pool, but self.config.project, that it still pools?)

File: fiftyone/utils/pe.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        self.layer_name = None
+
+        # defaults
+        # TODO: move to manifest
+        if self.entrypoint_fcn is None:
+            self.entrypoint_fcn = pe.VisionTransformer.from_config
+        if self.entrypoint_args is None:
+            self.entrypoint_args = {}
+        default_entrypoint_args = {
+            "name": "PE-Core-T16-384",
```
Comment (manushreegangwar) [inline fiftyone/utils/pe.py]: Could you define `DEFAULT_PE_MODEL = "PE-Core-T16-384"` and set this using that?

File: fiftyone/utils/pe.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        }
+        self.entrypoint_args = {
+            **default_entrypoint_args,
+            **self.entrypoint_args,
+        }
+        if "checkpoint_path" not in self.entrypoint_args:
+            self.entrypoint_args["checkpoint_path"] = (
+                fo.config.model_zoo_dir
+                + "/"
+                + self.entrypoint_args["name"]
```
Comment (manushreegangwar) [inline fiftyone/utils/pe.py]: If "name" is not in entrypoint args, this is throw an error

File: fiftyone/utils/pe.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+fou.ensure_package("perception_models")
+import perception_models.core.vision_encoder.pe as pe
+import perception_models.core.vision_encoder.transforms as pe_transforms
+
+
+def download_pe_checkpoint(name: str, checkpoint_path: str):
+    path = f"hf://facebook/{name}:{name}.pt"
+
+    # Load from huggingface
+    path = path[len("hf://") :]
```
Comment (manushreegangwar) [inline fiftyone/utils/pe.py]: "hf://" in `path` seems unneccesary

**Review by manushreegangwar — COMMENTED** [review]
Minor comments. Looks good otherwise 🚀 

@jacobsela Have you checked that documentation looks as expected for these models? Might need special handling.

File: fiftyone/utils/pe.py (hunk 4 above)
Comment (jacobsela) [reply fiftyone/utils/pe.py:129]: yes

Comment (jacobsela) [reply fiftyone/utils/pe.py:129]: need to pool to project. Technically we can project each token individually but not sure if that will give intended results, leaving it for future work.

File: fiftyone/utils/pe.py (hunk 3 above)
Comment (jacobsela) [reply fiftyone/utils/pe.py:88]: so that in applications users (like you) can just do `foz.load_zoo_model(PE-Core-L14-336-Vision-Encoder)` and get a model that behaves as intended.

Comment (kaixi-wang) [reply fiftyone/utils/pe.py:88]: Oh I wasn't questioning hardcoded defaults in general. I mean why are they in two places.

Comment (jacobsela) [reply fiftyone/utils/pe.py:88]: oh so that in the manifest we can just have minimal diffs from the defaults set here and so that if we want to change behavior across the board in the integration we can just do so here rather than for each entry in the manifest.

**Comment by jacobsela** [comment]
@manushreegangwar docs look fine

**Comment by manushreegangwar** [comment]
> @manushreegangwar docs look fine

@jacobsela for consistency, these should follow dino's [pattern](https://github.com/voxel51/fiftyone/blob/develop/docs/scripts/make_model_zoo_docs.py#L163C10-L163C14)
